engine/src/execution_graph/logic_controllers/taskflow/graph.cpp

```diff
@@ -0,0 +1,255 @@
+#include "graph.h"
+
+namespace ral {
+namespace cache {
+
+	kpair graph::operator+=(kpair p) {
+		std::string source_port_name = std::to_string(p.src->get_id());
+		std::string target_port_name = std::to_string(p.dst->get_id());
+
+		if(p.has_custom_source()) {
+			source_port_name = p.src_port_name;
+		}
+		if(p.has_custom_target()) {
+			target_port_name = p.dst_port_name;
+		}
+		this->add_edge(p.src, p.dst, source_port_name, target_port_name, p.cache_machine_config);
+		return p;
+	}
+
+	void graph::check_and_complete_work_flow() {
+		for(auto node : container_) {
+			kernel * kernel_node = node.second;
+			if(kernel_node) {
+				if(get_neighbours(kernel_node).size() == 0) {
+					Edge fake_edge = {.source = (int32_t) kernel_node->get_id(),
+						.target = -1,
+						.source_port_name = std::to_string(kernel_node->get_id()),
+						.target_port_name = ""};
+					edges_[kernel_node->get_id()].insert(fake_edge);
+				}
+			}
+		}
+	}
+
+	void graph::execute() {
+		check_and_complete_work_flow();
+
+		std::vector<BlazingThread> threads;
+		std::set<std::pair<size_t, size_t>> visited;
+		std::deque<size_t> Q;
+		for(auto start_node : get_neighbours(head_id_)) {
+			Q.push_back(start_node.target);
+		}
+		while(not Q.empty()) {
+			auto source_id = Q.front();
+			auto source = get_node(source_id);
+			auto source_edges = get_reverse_neighbours(source);
+			bool node_has_all_dependencies = source_edges.size() == 0 ? true : 
+				std::all_of(source_edges.begin(), source_edges.end(), [visited](Edge edge) { 
+					auto edge_id = std::make_pair(edge.source, edge.target);
+					return visited.find(edge_id) != visited.end();});
+			Q.pop_front();
+			if (node_has_all_dependencies){
+				if(source) {
+					for(auto edge : get_neighbours(source)) {
+						auto target_id = edge.target;
+						auto target = get_node(target_id);
+						auto edge_id = std::make_pair(source_id, target_id);
+						if(visited.find(edge_id) == visited.end()) {
+							visited.insert(edge_id);
+							Q.push_back(target_id);
+							BlazingThread t([this, source, source_id, edge] {
+								auto state = source->run();
+								if(state == kstatus::proceed) {
+									source->output_.finish();
+								} else if (edge.target != -1) { // not a dummy node
+									std::cout<<"ERROR kernel "<<source_id<<" did not finished successfully"<<std::endl;
+								}
+							});
+							threads.push_back(std::move(t));
+						} else {
+							// TODO: and circular graph is defined here. Report and error
+						}
+					}
+				}
+			} else { // if we dont have all the dependencies, lets put it back at the back and try it later
+				Q.push_back(source_id);
+			}
+		}
+		for(auto & thread : threads) {
+			thread.join();
+		}
+	}
+
+	void graph::show() {
+		check_and_complete_work_flow();
```

Comment:
ok its being used here, how is show used?

Reply:
see my comment above